batchlib/analysis/cell_analysis_qc.py

```diff
@@ -169,13 +208,135 @@ def run(self, input_files, output_files):
+                 cell_outlier_table_name='outliers',
+                 outlier_criteria=DEFAULT_WELL_OUTLIER_CRITERIA,
+                 **super_kwargs):
+
+        self.validate_outlier_criteria(outlier_criteria)
+        self.outlier_criteria = outlier_criteria
+
+        self.table_out_key = table_out_key
+        super().__init__(table_out_keys=[table_out_key],
+                         check_image_outputs=False,
+                         cell_seg_key=cell_seg_key,
+                         serum_key=serum_key,
+                         marker_key=marker_key,
+                         serum_bg_key=serum_bg_key,
+                         marker_bg_key=marker_bg_key,
+                         **super_kwargs)
+
+        output_group = cell_seg_key if self.identifier is None else cell_seg_key + '_' + self.identifier
+        self.cell_outlier_table = output_group + '/' + serum_key + '_' + cell_outlier_table_name
+
+    # we need to add this so that outlier cells are ignored when computing the ratios
+    def load_cell_outliers(self, input_file):
+        return load_cell_outlier_dict(input_file, self.cell_outlier_table, self.name)
+
+    def outlier_heuristics(self, in_files):
+        infected_stats, control_stats = self.load_per_cell_statistics(in_files)
+
+        n_images = len(in_files)
+        n_infected = len(infected_stats['labels'])
+        n_control = len(control_stats['labels'])
+        n_cells = n_infected + n_control
+
+        outlier_type = ''
+        is_outlier = 0
+
+        min_cells_per_im = self.outlier_criteria['min_number_cells_per_image']
+        if min_cells_per_im is not None and n_cells < min_cells_per_im * n_images:
+            is_outlier = 1
+            outlier_type += 'too few cells;'
+
+        max_cells_per_im = self.outlier_criteria['max_number_cells_per_image']
+        if max_cells_per_im is not None and n_cells < max_cells_per_im * n_images:
+            is_outlier = 1
+            outlier_type += 'too few cells;'
+
+        min_control_per_im = self.outlier_criteria['min_number_control_cells_per_image']
+        if min_control_per_im is not None and n_control < min_control_per_im * n_images:
+            is_outlier = 1
+            outlier_type += 'too few control cells;'
+
+        control_fraction = float(n_control) / n_cells
+        min_fraction = self.outlier_criteria['min_fraction_of_control_cells']
+        if min_fraction is not None and control_fraction < min_fraction:
+            is_outlier = 1
+            outlier_type += 'too small fraction of control cells;'
+
+        # check for negative ratios
+        if self.outlier_criteria['check_ratios']:
+            ratios = compute_ratios(control_stats, infected_stats, serum_key=self.serum_key)
+            for name, val in ratios.items():
+                if not name.startswith('ratio'):
+                    continue
+                if val < 0.:
+                    is_outlier = 1
+                    outlier_type += f'{name} is negative'
```

Comment:
You only do this for wells, correct? We have yet to add the image-level outlier detection to the well- and plate-wise background calculation. If this is only done for wells we can first calculate the feature tables per image, then detect the image outliers, then compute the bg values (ignoring outlier images) and only then do the well-level outlier control. If we also want to do this on an image level, it gets more complicated (But I do not see the need for it)

Reply:
Yes, I only do this for the wells. So let's keep it that way then we don't need to make it more complicated :).